wire-schema/src/main/java/com/squareup/wire/schema/IdentifierSet.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2015 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire.schema;
+
+import java.util.NavigableSet;
+import java.util.TreeSet;
+
+/**
+ * A heterogeneous set of type names and type members. If a member is added to the set, its type is
+ * implicitly also added. A type that is added without a specific member implicitly contains all
+ * of that type's members.
+ *
+ * <p>For example, an identifiers set populated with {@code Movie} and {@code Actor#name} contains
+ * all members of {@code Movie} (such as {@code Movie#name} and {@code Movie#release_date}). It
+ * contains the type {@code Actor} and one member {@code Actor#name}, but not {@code
+ * Actor#birth_date} or {@code Actor#oscar_count}.
+ */
+public final class IdentifierSet {
+  final NavigableSet<String> set = new TreeSet<>();
+
+  public boolean isEmpty() {
+    return set.isEmpty();
+  }
+
+  /** Adds a type name or member name. Returns true if the set was modified. */
```

Comment:
Calling `add("Foo#bar")` and then `add("Foo#baz")` will return true and then false, despite the set being modified the second time.